sdk-base/src/main/java/com/mapbox/maps/plugin/animation/MapboxAnimatorThread.kt

```diff
@@ -0,0 +1,184 @@
+    return (backgroundAnimatorThread ?: mainAnimatorThread).post(runnable).also { posted ->
+      if (!posted) {
+        logW(
+          TAG,
+          "post did not happen, ensure the app is in foreground. background  animator " +
+            "thread: ${backgroundAnimatorThread != null}"
+        )
+      }
+    }
+  }
+
+  /**
+   * Mapbox gl-native libraries require the main thread. Use this function whenever an animation
+   * task should be posted on the main thread, even when background threads are enabled.
+   *
+   * Run immediate if my looper is the designated looper, otherwise post the work to a handler
+   * queue. This is to reduce thread context switches while running animations.
+   */
+  fun postMain(runnable: Runnable): Boolean {
+    return mainAnimatorThread.post(runnable).also { posted ->
+      if (!posted) {
+        logW(TAG, "postMain did not happen, ensure the app is in foreground")
+      }
+    }
+  }
+
+  /**
+   * The beginning of the animator thread lifecycle. This will create resources for running
+   * animator threads.
+   */
+  fun onDetached() {
+    mainAnimatorThread.onDetached()
+    backgroundAnimatorThread?.onDetached()
+    backgroundAnimatorThread = null
+  }
+
+  private companion object {
+    private const val TAG = "MapboxAnimatorThread"
+  }
+}
+
+/**
+ * The [PackageManager] has issues being mocked. This class makes it so usages of the package
+ * meta data can be tested without mocking the application info metadata.
+ *
+ * https://github.com/mockk/mockk/issues/182
+ */
+@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+internal class AndroidManifestMetaData {
+
+  private fun metaData(): Bundle {
+    val context = MapboxSDKCommon.getContext()
+    val packageManager: PackageManager = context.packageManager
+    return packageManager.getApplicationInfo(
+      context.packageName,
+      PackageManager.GET_META_DATA
+    ).metaData
+  }
+
+  /**
+   * Perform manifest metadata lookup for boolean value of
+   * com.mapbox.maps.BackgroundAnimatorThread. Returns false by default.
+   */
+  fun backgroundAnimatorThreadEnabled(): Boolean {
+    return metaData().getBoolean(BACKGROUND_ANIMATOR_THREAD_KEY, false)
+  }
+
+  private companion object {
+    private const val BACKGROUND_ANIMATOR_THREAD_KEY = "com.mapbox.maps.BackgroundAnimatorThread"
+  }
+}
+
+private interface AnimatorThread {
+  @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+  val handler: Handler?
+
+  fun onAttached(): AnimatorThread
+  fun onDetached()
+
+  fun post(runnable: Runnable): Boolean {
+    return if (Looper.myLooper() == handler?.looper) {
+      runnable.run()
+      true
+    } else {
+      handler?.post(runnable) ?: false
+    }
+  }
+}
+
+private class BackgroundAnimatorThread : AnimatorThread {
+  private val threadName = "animation_thread_${UUID.randomUUID()}"
+  private var handlerThread: HandlerThread = HandlerThread(
+    threadName,
+    Process.THREAD_PRIORITY_DISPLAY
```

Comment:
why do we need such high priority? we already have render thread that uses such priority

Reply:
Yeah, could play with lower priority here. Just copy-paste from `MapboxRenderThread`

There are honestly many other big issues, creating high priority is mostly to make it easier to reproduce the other issues.

Reply:
Going to try `THREAD_PRIORITY_FOREGROUND`